## <b>■ 그룹연산</b>
    1. df.groupby
    2. group.함수()
    3. group.agg

### <b>■ 그룹객체.agg</b>
    문법
        group.agg(매핑함수)

    예제. 
```SQL
select deptno, sum(sal)
    from emp
    group by deptno;
```

In [3]:
import pandas as pd
emp = pd.read_csv('emp.csv')

df = emp[['sal','deptno']]

gp = df.groupby('deptno')
gp.agg(sum)

sal
deptno       
10       8750
20      10875
30       9400

### ※ 문제213. 아래의 SQL을 Pandas로 구현하시오
```sql
select deptno, job, sum(sal)
    from emp
    group by deptno, job
    order by deptno, job;
```

In [9]:
import pandas as pd
emp = pd.read_csv('emp.csv')
df = emp[['deptno','job','sal']]
gp = df.groupby(['deptno','job'])

gp.agg(sum).sort_values(by=['deptno','job'],ascending=True)

sal
deptno job            
10     CLERK      1300
       MANAGER    2450
       PRESIDENT  5000
20     ANALYST    6000
       CLERK      1900
       MANAGER    2975
30     CLERK       950
       MANAGER    2850
       SALESMAN   5600

### ※ 문제214. 아래의 SQL을 Pandas로 구현하시오
```sql
select deptno, min(sal), max(sal)
    from emp
    group by deptno;
```

In [10]:
import pandas as pd
emp = pd.read_csv('emp.csv')
df = emp[['deptno','sal']]
gp = df.groupby('deptno')
gp.agg([min,max])

sal      
         min   max
deptno            
10      1300  5000
20       800  3000
30       950  2850

### ※ 문제215. Kaggle 타이타닉의 클래스별 운임의 최대값과 최소값을 아래와 같이 출력하시오
            Fare
            min       max
    Pclass
    1       0.0       512.3292
    2       0.0        73.5000
    3       0.0        69.5500

In [12]:
import pandas as pd
tit = pd.read_csv('train.csv')
df = tit[['Pclass','Fare']]
gp = df.groupby(['Pclass'])
gp.agg([min,max])

Fare          
        min       max
Pclass               
1       0.0  512.3292
2       0.0   73.5000
3       0.0   69.5500

## <b>■ 그룹연산</b>
    1. df.groupby
    2. group.함수()
    3. group.agg
    4. group.filter(조건식 함수)
    
### <b>■ group.filter(조건식 함수)</b>
    그룹객체에 filter() 메소드를 적용할 때 조건식을 가진 함수를 전달하면 조건이 참인 그룹만을 남긴다.
    문법
        group.filter(조건식 함수)

##### 예제. 부서번호별 평균월급이 2000보다 작은 사원들의 이름, 월급, 직업, 부서번호를 출력
```sql
select ename, sal, job, 부서평균
    from (
            select ename, sal, job, deptno, avg(sal) over (partition by deptno) 부서평균
                from emp
         )
    where 부서평균 < 2000;
```

In [18]:
import pandas as pd
emp = pd.read_csv('emp.csv')
df = emp[['ename','sal','job','deptno']]
gp = df.groupby('deptno')
gp.filter(lambda x:x.sal.mean()<2000)

ename   sal       job  deptno
1   BLAKE  2850   MANAGER      30
4  MARTIN  1250  SALESMAN      30
5   ALLEN  1600  SALESMAN      30
6  TURNER  1500  SALESMAN      30
7   JAMES   950     CLERK      30
8    WARD  1250  SALESMAN      30

## <b>■ 그룹연산</b>
    1. df.groupby
    2. group.함수()
    3. group.agg
    4. group.filter(조건식 함수)
    5. group.apply(매핑 함수)
    
### <b>■ group.apply</b>
    apply() 메소드는 Pandas 객체의 개별원소를 특정함수에 일대일매핑
    문법
        group.apply(매핑함수)

In [22]:
import pandas as pd
tit = pd.read_csv('train.csv')
df = tit[['Name','Survived','Age','Pclass','Fare']]
gp = df.groupby('Pclass')
rs = gp.apply(lambda x:x.describe())
rs

Survived         Age  Pclass        Fare
Pclass                                                  
1      count  216.000000  186.000000   216.0  216.000000
       mean     0.629630   38.233441     1.0   84.154687
       std      0.484026   14.802856     0.0   78.380373
       min      0.000000    0.920000     1.0    0.000000
       25%      0.000000   27.000000     1.0   30.923950
       50%      1.000000   37.000000     1.0   60.287500
       75%      1.000000   49.000000     1.0   93.500000
       max      1.000000   80.000000     1.0  512.329200
2      count  184.000000  173.000000   184.0  184.000000
       mean     0.472826   29.877630     2.0   20.662183
       std      0.500623   14.001077     0.0   13.417399
       min      0.000000    0.670000     2.0    0.000000
       25%      0.000000   23.000000     2.0   13.000000
       50%      0.000000   29.000000     2.0   14.250000
       75%      1.000000   36.000000     2.0   26.000000
       max      1.000000   70.000000     2.0   73.500000
3      count  491.000000  355.000000   491.0  491.000000
       mean     0.242363   25.140620     3.0   13.675550
       std      0.428949   12.495398     0.0   11.778142
       min      0.000000    0.420000     3.0    0.000000
       25%      0.000000   18.000000     3.0    7.750000
       50%      0.000000   24.000000     3.0    8.050000
       75%      0.000000   32.000000     3.0   15.500000
       max      1.000000   74.000000     3.0   69.550000

    클래스별로 생존, 나이, 운임에 대한 건수, 평균값, 최대값, 최소값, 사분위수 값, 표준편차를 출력
    
    머신러닝 알고리즘으로 기계를 학습시킬 때 변수의 단위가 다르면 학습이 잘 안될 수 있으므로 동일한 수치 단위로 조정 필요
        -> 정규화
    
#### <b>면접질문</b>
    1. 데이터의 종류에 따라 적합한 머신러닝 알고리즘이 무엇인가
    2. 회귀분석에서 결정계수가 무엇인가
    3. 데이터 정규화를 하는 이유가 무엇인가
        데이터가 가진 feature의 스케일이 심하게 차이가 나는 경우에는 모델이 학습할 때 학습이 잘 안될 수 있으므로 
        모든 데이터의 포인트가 동일한 스케일(중요도)로 반영되도록 해주기 위해서
    4. 데이터의 정규화의 방법에는 무엇이 있나
        1) min-max 정규화(최소-최대 정규화)
$${{x-min(x)}\over{max(x)-min(x)}}$$

            모든 feature들의 스케일이 동일해지지만 이상치(outlier)를 잘 처리하지 못한다.
            
        2) z-score 정규화(Z-점수 정규화)
$${{x- \mu}\over {\sigma}}$$

            이상치(outlier)는 잘 처리하지만, 정확히 동일한 척도로 정규화된 데이터를 생성하지 않는다.
            
#### ※ 생각해야할 문제. 이상치(outlier)를 잘 처리한다는 것은 무슨 얘기인가?

In [25]:
# 예제. z-score 정규화
import pandas as pd
tit = pd.read_csv('train.csv')
df = tit[['Name','Survived','Age','Pclass','Fare']]
gp = df.groupby('Pclass')
def z_score(x):
    return ((x-x.mean())/x.std())
agg_gp = gp.Age.apply(z_score)
agg_gp, agg_gp.describe()

(0     -0.251342
 1     -0.015770
 2      0.068776
 3     -0.218434
 4      0.789041
          ...   
 886   -0.205529
 887   -1.299306
 888         NaN
 889   -0.826424
 890    0.548953
 Name: Age, Length: 891, dtype: float64,
 count    7.140000e+02
 mean    -7.945714e-17
 std      9.985965e-01
 min     -2.520692e+00
 25%     -5.714600e-01
 50%     -6.268304e-02
 75%      5.922208e-01
 max      3.910190e+00
 Name: Age, dtype: float64)

### ※ 문제216. Kaggle 타이타닉 데이터의 운임을 min-max 정규화 하시오

In [26]:
import pandas as pd
tit = pd.read_csv('train.csv')
df = tit[['Name','Survived','Age','Pclass','Fare']]
gp = df.groupby('Pclass')
def min_max_norm(x):
    return((x-min(x))/(max(x)-min(x)))
gp.Fare.apply(min_max_norm)

0      0.104242
1      0.139136
2      0.113947
3      0.103644
4      0.115744
         ...   
886    0.176871
887    0.058556
888    0.337168
889    0.058556
890    0.111431
Name: Fare, Length: 891, dtype: float64

## <b>■ 그룹연산</b>
    1. df.groupby
    2. group.함수()
    3. group.agg
    4. group.filter(조건식 함수)
    5. group.apply(매핑 함수)
    6. 멀티인덱스
    
### <b>■ 멀티인덱스</b>
    groupby() 메소드에 여러 열을 리스트 형태로 전달하면 각 열들이 다중으로 행 인덱스를 구성하는 것을 확인했는데 
    이처럼 Pandas는 행 인덱스를 여러 레벨로 구현할 수 있도록 멀티인덱스 클래스를 지원

In [28]:
import pandas as pd
tit = pd.read_csv('train.csv')
df = tit[['Name','Survived','Age','Pclass','Fare','Sex']]
gp = df.groupby(['Pclass','Sex'])
gdf = gp.mean()
gdf

Survived        Age        Fare
Pclass Sex                                    
1      female  0.968085  34.611765  106.125798
       male    0.368852  41.281386   67.226127
2      female  0.921053  28.722973   21.970121
       male    0.157407  30.740707   19.741782
3      female  0.500000  21.750000   16.118810
       male    0.135447  26.507589   12.661633

### ※ 문제217. 위의 결과에서 first class만 따로 검색하시오

In [31]:
import pandas as pd
tit = pd.read_csv('train.csv')
df = tit[['Name','Survived','Age','Pclass','Fare','Sex']]
gp = df.groupby(['Pclass','Sex'])
gdf = gp.mean()
gdf.loc[1]

Survived        Age        Fare
Sex                                    
female  0.968085  34.611765  106.125798
male    0.368852  41.281386   67.226127

### ※ 문제218. 위의 결과에서 first클래스의 female 데이터만 검색하시오

In [36]:
import pandas as pd
tit = pd.read_csv('train.csv')
df = tit[['Name','Survived','Age','Pclass','Fare','Sex']]
gp = df.groupby(['Pclass','Sex'])
gdf = gp.mean()
gdf.loc[1,'female']

Survived      0.968085
Age          34.611765
Fare        106.125798
Name: (1, female), dtype: float64